// @generated SignedSource<<61bc2c31915f16ea635205b1d9ce6dc6>>
// DO NOT EDIT THIS FILE MANUALLY!
// This file is a mechanical copy of the version in the configerator repo. To
// modify it, edit the copy in the configerator repo instead and copy it over by
// running the following in your fbcode directory:
//
// configerator-thrift-updater scm/mononoke/adaptive_rate_limiter/adaptive_rate_limiter.thrift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

package "meta.com/scm/mononoke/adaptive_rate_limiter"

namespace py3 scm.mononoke.adaptive_rate_limiter
namespace py scm.mononoke.adaptive_rate_limiter.adaptive_rate_limiter

// Adaptive Rate Limiter Configuration
// This defines configuration for CPU and memory based adaptive rate limiting
// that can be used across Mononoke services

/// Operation mode for adaptive rate limiting
enum OperationMode {
  /// Rate limiter is disabled - no monitoring or shedding
  DISABLED = 0,
  /// Rate limiter is enabled - actively shed requests when limits are exceeded
  ENABLED = 1,
  /// Dry run mode - monitor and log what would be shed, but don't actually shed requests
  DRY_RUN = 2,
}

/// Resource monitoring mode for adaptive rate limiting
enum AdaptiveRateLimiterMode {
  /// No resource monitoring - rate limiter disabled
  NONE = 0,
  /// Monitor only cgroup (container) resources
  CGROUP_ONLY = 1,
  /// Monitor only host-level resources
  HOST_ONLY = 2,
  /// Monitor both cgroup and host resources
  BOTH = 3,
}

/// Configuration for adaptive rate limiter based on CPU and memory thresholds
struct AdaptiveRateLimiterConfig {
  /// Operation mode - controls whether rate limiter is enabled, disabled, or in dry run
  1: OperationMode operation_mode = OperationMode.DISABLED;
  /// Resource monitoring mode
  2: AdaptiveRateLimiterMode monitoring_mode = AdaptiveRateLimiterMode.NONE;
  /// CPU soft limit ratio (0.0 to 1.0) - start probabilistic shedding
  3: double cpu_soft_limit_ratio = 0.90;
  /// CPU hard limit ratio (0.0 to 1.0) - aggressive shedding
  4: double cpu_hard_limit_ratio = 0.95;
  /// Memory soft limit ratio (0.0 to 1.0) - start probabilistic shedding
  5: double mem_soft_limit_ratio = 0.8;
  /// Memory hard limit ratio (0.0 to 1.0) - aggressive shedding
  6: double mem_hard_limit_ratio = 0.85;
  /// Load update period in milliseconds
  7: i64 load_update_period_ms = 100;
}
